spec/cedar_policy/entities_spec.rb

```diff
@@ -37,4 +37,36 @@
       it { is_expected.to be_one }
     end
   end
+
+  describe "with schema" do
+    let(:schema) do
+      CedarPolicy::Schema.new(
+        <<~SCHEMA
+          entity User {
+            isAdmin?: Bool
+          };
+
+          entity Image {
+            owner: User
+          };
+
+          action View  appliesTo {
+              principal: [User],
+              resource: [Image]
+          };
+        SCHEMA
+      )
+    end
+
+    let(:user) { {uid: {type: "User", id: "1"}, attrs: {}, parents: []} }
+    let(:admin) { {uid: {type: "User", id: "2"}, attrs: {isAdmin: true}, parents: []} }
+    let(:image) { {uid: {type: "Image", id: "1"}, attrs: {owner: {type: "User", id: 1}}, parents: []} }
+
+    subject(:entities) do
+      CedarPolicy::Entities.new([user, admin, image], schema: schema)
+    end
+
+    it { is_expected.to(satisfy { |entities| entities.to_ary.size == 3 }) }
+    it { is_expected.to(have_attributes(schema: schema)) }
```

Comment:
I have a question: Is the `schema` comparable or not?

1. `Schema.new('ABC') == Schema.new('ABC')`
2. `Schema.new('ABC') != Schema.new('ABC')`

If it can be compared, I prefer not to use `it { is_expected.to(have_attributes(schema: schema)) }`
The reason is assertation should not use a reference e.g. `it { is_expected.to(have_attributes(schema: Schema. new('ABC'))) }` and the reader can know which value to compare and it should be the same.

Reply:
I don't usually use rspec for ruby tests, so I'm not as familiar with typical rspec usage. I apologize if I'm misunderstanding what you are saying, but the short answer is that `Schema.new('entity ABC;') != Schema.new('entity ABC;')` — Schema objects cannot be compared by value.

At the rust level, `cedar_policy::Schema` is not comparable as far as I understand. It doesn't implement `Eq` or `PartialEq`. Since `RScema` is just a simple wrapper struct for `cedar_policy::Schema`, it doesn't support comparisons either. So the question of whether schema is comparable or not is just a ruby question.

When it comes to the ruby layer, ruby doesn't know anything about what is inside of the rust object, and so all it has to work with for testing equality is the basic implementation of the `==` operator provided by ruby's built-in `BasicObject`, which ultimately just checks if both sides have the same `object_id`. That is actually what I wanted to test in this case, because I want to assert that the Schema I passed to the Entities constructor is now stored as the `schema` attribute.

If we could validate that the schema was being _used to parse the entities_, that would be a stronger test (we could feed it some non-conforming entities data and check for an appropriate exception), but there's nothing in the Entities object's ruby API that triggers that parsing right now (it happens in `try_convert`, which is triggered at the last moment inside of the authorizer, I think). So for now, I _think_ the best we can do is validate that the Entities object "has" the Schema object we passed to the constructor. If you have more ideas, though, I am happy to hear them.